lib/matplotlib/category.py

```diff
@@ -1,128 +1,109 @@
+        majfmt = StrCategoryFormatter(axis.units)
         return units.AxisInfo(majloc=majloc, majfmt=majfmt)
 
     @staticmethod
     def default_units(data, axis):
-        # the conversion call stack is:
-        # default_units->axis_info->convert
-        if axis.unit_data is None:
-            axis.unit_data = UnitData(data)
-        else:
-            axis.unit_data.update(data)
-        return None
+        return UnitData()
 
 
-class StrCategoryLocator(ticker.FixedLocator):
-    def __init__(self, locs):
-        self.locs = locs
-        self.nbins = None
+class StrCategoryLocator(ticker.Locator):
+    def __init__(self, unit_data):
+        self._unit_data = unit_data
 
+    def __call__(self):
+        return list(self._unit_data._mapping.values())
 
-class StrCategoryFormatter(ticker.FixedFormatter):
-    def __init__(self, seq):
-        self.seq = seq
-        self.offset_string = ''
+    def tick_values(self, vmin, vmax):
+        return self()
 
 
-class UnitData(object):
-    # debatable makes sense to special code missing values
-    spdict = {'nan': -1.0, 'inf': -2.0, '-inf': -3.0}
+class StrCategoryFormatter(ticker.Formatter):
+    def __init__(self, unit_data):
+        self._unit_data = unit_data
+
+    def __call__(self, x, pos=None):
+        if pos is None:
+            return ""
+        r_mapping = {v: k for k, v in self._unit_data._mapping.items()}
+        return r_mapping.get(int(x), '')
 
-    def __init__(self, data):
-        """Create mapping between unique categorical values
-        and numerical identifier
+
+class UnitData(object):
+    def __init__(self, data=None):
+        """Create mapping between unique categorical values and numerical id.
 
         Parameters
         ----------
-        data: iterable
-            sequence of values
+        data : Mapping[str, int]
+            The initial categories.  May be `None`.
+
         """
-        self.seq, self.locs = [], []
-        self._set_seq_locs(data, 0)
-
-    def update(self, new_data):
-        # so as not to conflict with spdict
-        value = max(max(self.locs) + 1, 0)
-        self._set_seq_locs(new_data, value)
-
-    def _set_seq_locs(self, data, value):
-        strdata = shim_array(data)
-        new_s = [d for d in np.unique(strdata) if d not in self.seq]
-        for ns in new_s:
-            self.seq.append(ns)
-            if ns in UnitData.spdict:
-                self.locs.append(UnitData.spdict[ns])
-            else:
-                self.locs.append(value)
-                value += 1
+        self._vals = []
+        if data is None:
+            data = ()
+        self._mapping = OrderedDict(data)
+        for k, v in self._mapping.items():
+            if not isinstance(k, six.text_type):
+                raise TypeError("{val!r} is not a string".format(val=k))
+            self._mapping[k] = int(v)
+        if self._mapping:
+            start = max(self._mapping.values()) + 1
```

Comment:
I feel like if you pass an explicit mapping then either
1. you don't expect new categories to come in and you'd rather error out on new categories, or
2. you may want to have your own index-generator as well.

thoughts?